smart_card_connector_app/src/chrome-api-jstocxxtest.js

```diff
@@ -867,6 +867,34 @@ goog.exportSymbol('testChromeApiProviderToCpp', {
       // The test `tearDown()` will verify that the disconnection works despite
       // the unended transaction.
     },
+
+    // Test that EndTransaction succeeds after transaction has begun.
+    'testEndTransaction_success': async function() {
+      // Start transaction
+      expectReportPlainResult(/*requestId=*/ 111, 'SUCCESS');
+      await mockChromeApi
+          .dispatchEvent(
+              'onBeginTransactionRequested', /*requestId=*/ 111, readerHandle)
+          .$waitAndVerify();
+
+      expectReportPlainResult(/*requestId=*/ 112, 'SUCCESS');
+      await mockChromeApi
+          .dispatchEvent(
+              'onEndTransactionRequested', /*requestId=*/ 112, readerHandle,
+              chrome.smartCardProviderPrivate.Disposition.LEAVE_CARD)
+          .$waitAndVerify();
+    },
+
+
+    // Test that EndTransaction fails without BeginTransaction.
+    'testEndTransaction_fail': async function() {
+      expectReportPlainResult(/*requestId=*/ 111, 'NOT_TRANSACTED');
```

Comment:
As discussed offline, it's a bit of a weird error code, but maybe that's what PC/SC-Lite always been doing. I'll add a similar C++ test (seems good anyway) and try to track down the code.

Reply:
thanks, let's do this :)